code/modules/projectiles/guns/projectile/constructable/siegecannon.dm

```diff
@@ -0,0 +1,221 @@
+		return
+	if(!isliving(C))
+		return
+	if(loadedMob || loadedItem)
+		to_chat(user,"<span class='warning'>The [src] is already full.</span>" )
+		return
+	visible_message("<span class='warning'>[user] is stuffing [C] into \the [src].</span>")
+	if(do_after(user, C, 3 SECONDS))
+		loadMob(C, user)
+
+/obj/structure/siege_cannon/proc/loadMob(var/mob/living/mLoad, mob/user)
+	mLoad.forceMove(src)
+	loadedMob = mLoad
+
+/obj/structure/siege_cannon/relaymove(mob/user)
+	if(do_after(user, src, 1 SECONDS))
+		if(user.loc == src)
+			user.forceMove(loc)
+		loadedMob = null
+
+/obj/structure/siege_cannon/proc/fireCheck()
+	if(!wFuel)
+		return FALSE
+	if(!loadedItem && !loadedMob)
+		return FALSE
+	return TRUE
+
+/obj/structure/siege_cannon/proc/siegeFire(mob/user)
+	if(!fireCheck())
+		return
+	spark(src)
+	playsound(src, 'sound/effects/Explosion_Small1.ogg', 100, 1)
+	if(loadedItem)
+		itemFire()
+	else if(loadedMob)
+		mobFire()
+	wFuel = 0
+
+/obj/structure/siege_cannon/proc/itemFire()
+	var/atom/target = get_edge_target_turf(src, dir)
+	if(istype(loadedItem, /obj/item/cannonball) && wFuel == maxFuel)
+		var/obj/item/cannonball/CB = loadedItem
+		CB.cannonAdjust()
+	var/fireSpeed = calcSpeed()
+	loadedItem.forceMove(loc)
+	loadedItem.throw_at(target, wFuel*2, fireSpeed)
+	loadedItem = null
+
+/obj/structure/siege_cannon/proc/calcSpeed()
+	var/FS = (wFuel/2)*(round(loadedItem.w_class/2, 1))
+	return FS
+
+/obj/structure/siege_cannon/proc/mobFire()
+	var/atom/target = get_edge_target_turf(src, dir)
+	var/mSpeed = 32		//Twice that of an non-emagged mass driver.
+	if(M_FAT in loadedMob.mutations)
+		mSpeed = 40	//Big boi
+	loadedMob.forceMove(loc)
+	loadedMob.throw_at(target, wFuel*2, mSpeed)
+	loadedMob = null
+
+/obj/structure/siege_cannon/verb/rotate_cw()
+	set name = "Rotate (Clockwise)"
+	set category = "Object"
+	set src in oview(1)
+
+	src.dir = turn(src.dir, -90)
+	return 1
+
+/obj/structure/siege_cannon/verb/rotate_ccw()
+	set name = "Rotate (Counter-Clockwise)"
+	set category = "Object"
+	set src in oview(1)
+
+	src.dir = turn(src.dir, 90)
+	return 1
+
+
+//Cannonball////////
+
+/obj/item/cannonball
+	name = "cannonball"
+	desc = "A large sphere of iron."
+	icon = 'icons/obj/siege_cannon.dmi'
+	icon_state = "cannonball"
+	throwforce = 3
+	throw_speed = 1
+	throw_range = 2	//Rather heavy
+	force = 10	//Somehow less than a toolbox but okay
+	starting_materials = list(MAT_IRON = CC_PER_SHEET_METAL*3)
+	w_type = RECYK_METAL
+	flags = FPRINT | TWOHANDABLE | MUSTTWOHAND
+	var/cannonFired = FALSE
+
+/obj/item/cannonball/proc/cannonAdjust()
+	if(!cannonFired)
+		cannonFired = TRUE
+		throw_range = 50
+		throw_speed = 4
+		throwforce = 15
+	else
+		cannonFired = FALSE
+		throw_range = 2
+		throw_speed = 1
+		throwforce = 3
+
+/obj/item/cannonball/throw_impact(atom/hit_atom)
+	..()
+	if(cannonFired && !hit_atom.gcDestroyed)
+		cannonEffect(hit_atom)
+
+/obj/item/cannonball/proc/cannonEffect(var/atom/cTarg)
```

Comment:
This is such a crime against OOP that I cannot in good faith support it.
Atoms already have an `hitby` proc which seems like a more natural place to put this.